controller/src/resource_controller/action.rs

```diff
@@ -151,8 +182,52 @@ async fn creation_completed_action(r: &ResourceInterface) -> Result<CreationActi
     }
 }
 
+/// Destruction for a resource may be required if the resources `DestructionPolicy` is
+/// `OnTestSuccess` or `OnTestCompletion`.
+async fn deletion_needed(r: &ResourceInterface) -> Result<bool> {
+    let destruction_policy = r.resource().spec.destruction_policy;
+    if matches!(
+        destruction_policy,
+        DestructionPolicy::OnTestCompletion | DestructionPolicy::OnTestSuccess
+    ) {
```

Comment:
I prefer a check to early return here so we can un-indent most of this logic so it's not as nested. The check on resource dependencies after this is also independent of the destruction policy.